fix: Fix GitHub pull request mergeability for multiple required workflow runs

ajax-ryzhyi-r opened the pull request:
## what
This PR modifies the mergeability evaluation logic for GitHub pull requests by evaluating the latest workflow check run (by comparing `runNumber`'s) instead of the first one returned from the GitHub API.
<!--
- Describe high-level what changed as a result of these commits (i.e. in plain-english, what do these changes mean?)
- Use bullet points to be concise and to the point.
-->


## why
There is an issue described in detail in #5048 where Atlantis treats a PR as unmergeable when there are multiple required GiHub Actions workflow runs for the last commit of the branch and the first of them is failed atlantis responds with the error `Apply Failed: Pull request must be mergeable before running apply.` when attempting to use the `atlantis apply` command.

Upon investigating this issue, I discovered that GitHub returns all `checkRuns` in chronological order when Atlantis retrieves a pull request's last commit mergeability status. However, Atlantis iterates through these `checkRuns` and retrieves the first `checkRun` instead of the last one, which represents the current status.

<!--
- Provide the justifications for the changes (e.g. business case). 
- Describe why these changes were made (e.g. why do these commits fix the problem?)
- Use bullet points to be concise and to the point.
-->

## tests

<!--
- [ ] I have tested my changes by ...
-->
- I have tested my changes by running this fix in our Atlantis setup and everything works as expected
- I have updated automated tests to cover these cases

## references

<!--
- Link to any supporting github issues or helpful documentation to add some context (e.g. stackoverflow). 
- Use `closes #123`, if this PR closes a GitHub issue `#123`
-->
Fixes the issue discussed in comments of #5048

henriklundstrom commented on server/events/vcs/github_client.go:
```diff
@@ -719,9 +719,11 @@ func StatusContextPassed(statusContext StatusContext, vcsstatusname string) bool
-	for _, checkRun := range checkRuns {
-		if checkRun.Name == expectedContext {
-			return CheckRunPassed(checkRun)
+	// Iterate through checkRuns from the end, as GitHub returns them in chronological order
```
I'm not sure the order can be relied upon. Is it documented that it shall be chronological?

Alternatively, there is the `checkSuite.workflowRun.runNumber` that could be added to the query and used to find the latest check run for the given check.

ajax-ryzhyi-r replied:
> I'm not sure the order can be relied upon. Is it documented that it shall be chronological?
> 
> Alternatively, there is the `checkSuite.workflowRun.runNumber` that could be added to the query and used to find the latest check run for the given check.

I couldn't find any mentions of order in the documentation; this observation is based solely on empirical evidence.

Thank you for the suggestion. It looks more reliable, and I'll try to implement it 🤝

ajax-ryzhyi-r replied:
> I'm not sure the order can be relied upon. Is it documented that it shall be chronological?
> 
> Alternatively, there is the `checkSuite.workflowRun.runNumber` that could be added to the query and used to find the latest check run for the given check.

@henriklundstrom  I've rewritten the fix to use `checkSuite.workflowRun.runNumber` for determining the latest check run. Take a look, please.

X-Guardian commented:
Hi @ajax-ryzhyi-r, you have some spurious file changes in this PR that aren't relevant to your change. Can you revert them please? (Adding EOF LFs t files you haven't changed.)

henriklundstrom commented on server/events/vcs/github_client.go:
```diff
@@ -709,6 +710,24 @@ pagination:
+		}
+	}
+
+	for _, checkRun := range checkRuns {
```
There is no need to iterate twice, the check run with the highest run number can be returned immediately after the first loop if it is kept as a variable pointer.

nitrocode commented:
cc @ajax-ryzhyi-r if you can address the above comments, this pr has a higher chance of getting into the next release which should be fairly soon

ajax-ryzhyi-r commented:
> cc @ajax-ryzhyi-r if you can address the above comments, this pr has a higher chance of getting into the next release which should be fairly soon

@nitrocode I've already addressed all the comments above. If there's anything else I need to fix, please let me know

lukemassa commented:
> > cc @ajax-ryzhyi-r if you can address the above comments, this pr has a higher chance of getting into the next release which should be fairly soon
> 
> @nitrocode I've already addressed all the comments above. If there's anything else I need to fix, please let me know

If you could click "Resolved" for the issues that you've addressed just to confirm there's nothing left.

It looks like there might still be some end-of-line issues, for example https://github.com/runatlantis/atlantis/pull/5057/files#diff-adec0aa1d9d8bb750060e06ba970ea7ae5817c8517d94a862da133a95e1d2466R95, if you could take a look?

Otherwise the logic makes sense to me, I'm going to see if I can replicate the issue and confirm the new code fix

ajax-ryzhyi-r commented:
> > > cc @ajax-ryzhyi-r if you can address the above comments, this pr has a higher chance of getting into the next release which should be fairly soon
> > 
> > 
> > @nitrocode I've already addressed all the comments above. If there's anything else I need to fix, please let me know
> 
> If you could click "Resolved" for the issues that you've addressed just to confirm there's nothing left.
> 
> It looks like there might still be some end-of-line issues, for example https://github.com/runatlantis/atlantis/pull/5057/files#diff-adec0aa1d9d8bb750060e06ba970ea7ae5817c8517d94a862da133a95e1d2466R95, if you could take a look?
> 
> Otherwise the logic makes sense to me, I'm going to see if I can replicate the issue and confirm the new code fix

I cannot hide comments with "Resolved" status which I don't own ;)

EOF related changes in the files I've edited appear to be legitimate due to the .editorconfig settings in this repository -https://github.com/runatlantis/atlantis/blob/main/.editorconfig#L7

henriklundstrom commented on server/events/vcs/github_client.go:
```diff
@@ -718,10 +719,27 @@ func StatusContextPassed(statusContext StatusContext, vcsstatusname string) bool
 }
 
 func ExpectedCheckPassed(expectedContext githubv4.String, checkRuns []CheckRun, statusContexts []StatusContext, vcsstatusname string) bool {
+	// If there's no WorkflowRun, we assume there's only one CheckRun with the given name.
+	// In this case, we evaluate and return the status of this CheckRun.
```
This ought to be investigated further, but since the behaviour remains unchanged for this case, it shouldn't block this pull request. I might look into it at some point if I have the time.

X-Guardian commented:
@henriklundstrom, can you resolve the conversations that you started if you happy with the responses, and then we can get this PR merged.

henriklundstrom commented:
> @henriklundstrom, can you resolve the conversations that you started if you happy with the responses, and then we can get this PR merged.

@X-Guardian I actually can not! But I can say here that I'm happy with all of them, and let's proceed.

X-Guardian commented:
Thanks for your work on this @ajax-ryzhyi-r. Can you test it using one of the following dev images: [dev-debian-b975301](https://github.com/orgs/runatlantis/packages/container/atlantis/308188598?tag=dev-debian-b975301) or [dev-alpine-b975301](https://github.com/orgs/runatlantis/packages/container/atlantis/308188520?tag=dev-alpine-b975301)
